File: aspnet-in-practice/15/VB/VirtualPathProvider/App_Code/HashCodeCombiner.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Web
Imports System.Reflection

' wrapping class for System.Web.Util.HashCodeCombiner
Namespace ASPNET4InPractice

	Public Class HashCodeCombiner
		Private internalCombiner As Object
		Private Shared constructor As ConstructorInfo
		Private Shared addObjectMethod As MethodInfo
		Private Shared resultProperty As PropertyInfo

		Private Const flags As BindingFlags = BindingFlags.NonPublic Or BindingFlags.[Public] Or BindingFlags.Instance

		Shared Sub New()
			Dim asm As Assembly = Assembly.GetAssembly(GetType(System.Web.UI.Page))
			Dim hashCodeCombinerType As Type = asm.[GetType]("System.Web.Util.HashCodeCombiner")
			constructor = hashCodeCombinerType.GetConstructor(flags, Nothing, Type.EmptyTypes, Nothing)
			addObjectMethod = hashCodeCombinerType.GetMethod("AddObject", flags, Nothing, New Type() {GetType(Object)}, Nothing)
			resultProperty = hashCodeCombinerType.GetProperty("CombinedHashString", flags)
		End Sub

		Public Sub New()
			internalCombiner = constructor.Invoke(Nothing)
		End Sub

		Public Sub AddObject(ByVal value As Object)
			addObjectMethod.Invoke(internalCombiner, New Object() {value})
		End Sub

		Public ReadOnly Property CombinedHashString() As String
			Get
				Return DirectCast(resultProperty.GetValue(internalCombiner, Nothing), String)
			End Get
		End Property
	End Class
End Namespace

File: aspnet-in-practice/15/CS/VirtualPathProvider/App_Code/HashCodeCombiner.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Reflection;

// wrapping class for System.Web.Util.HashCodeCombiner
namespace ASPNET4InPractice
{

	public class HashCodeCombiner
	{
		private object internalCombiner;
		private static ConstructorInfo constructor;
		private static MethodInfo addObject;
		private static PropertyInfo resultProperty;

		private const BindingFlags flags = BindingFlags.NonPublic |
										   BindingFlags.Public |
										   BindingFlags.Instance;

		static HashCodeCombiner()
		{
			Assembly asm = Assembly.GetAssembly(typeof(System.Web.UI.Page));
			Type hashCodeCombinerType = asm.GetType("System.Web.Util.HashCodeCombiner");
			constructor = hashCodeCombinerType.GetConstructor(flags, null, Type.EmptyTypes, null);
			addObject = hashCodeCombinerType.GetMethod("AddObject", flags, null, new Type[] { typeof(object) }, null);
			resultProperty = hashCodeCombinerType.GetProperty("CombinedHashString", flags);
		}

		public HashCodeCombiner()
		{
			internalCombiner = constructor.Invoke(null);
		}

		public void AddObject(object value)
		{
			addObject.Invoke(internalCombiner, new object[] { value });
		}

		public string CombinedHashString
		{
			get
			{
				return (string)resultProperty.GetValue(internalCombiner, null);
			}
		}
	}
}